padl/util_transforms.py

```diff
@@ -52,29 +50,39 @@ def __call__(self, args):
             return self.if_.pd_call_transform(args)
         return self.else_.pd_call_transform(args)
 
-    def _pd_preprocess_part(self):
-        pre = type(self)(
-            if_=self.if_.pd_preprocess,
-            target_stage=self.target_stage,
-            else_=self.else_.pd_preprocess,
-        )
-        return pre
-
-    def _pd_forward_part(self):
-        forward = type(self)(
-            if_=self.if_.pd_forward,
-            target_stage=self.target_stage,
-            else_=self.else_.pd_forward,
-        )
-        return forward
-
-    def _pd_postprocess_part(self):
-        post = type(self)(
-            if_=self.if_.pd_postprocess,
-            target_stage=self.target_stage,
-            else_=self.else_.pd_postprocess,
-        )
-        return post
+    def _pd_get_splits(self, input_components=0):
+        if self._pd_splits is None or self._pd_splits[0][0] != input_components:
+            transforms = [self.if_, self.else_]
+            splits = ([], [], [])
+            # we need one component info per sub-transform - if it's not a list that means
+            # all are the same - we make it a list
+            input_components_ = input_components
+            if not isinstance(input_components_, list):
+                input_components_ = [input_components for _ in range(len(transforms))]
+
+            # go through the sub-transforms ...
+            output_components = []
+            for transform_, input_component in zip(transforms, input_components_):
+                (_, sub_output_components), subsplits = transform_._pd_get_splits(input_component)
+                output_components.append(sub_output_components)
+                for split, subsplit in zip(splits, subsplits):
+                    split.append(subsplit)
+
+            cleaned_splits = tuple(
+                builtin_identity if all(isinstance(s, Identity) for s in split) else split
+                for split in splits
+            )
+
+            final_splits = tuple(
+                IfInStage(
```

Comment:
@wuhu I'm not able to create a new instance of `IfInStage` in `_pd_get_splits`. I get the following error message when running `pytest -k TestIfInStage`
```python
    def test_infer_apply(self):
>       assert self.transform_1.infer_apply(1) == 9

tests/unittests/test_util_transforms.py:45: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
padl/transforms.py:734: in infer_apply
    self.pd_forward_device_check()
padl/transforms.py:1222: in pd_forward_device_check
    if isinstance(self.pd_forward, type(self)):
padl/transforms.py:639: in pd_forward
    forward = self._pd_get_splits()[1][1]
padl/transforms.py:1333: in _pd_get_splits
    (_, output_components), subsplits = transform_._pd_get_splits(output_components)
padl/util_transforms.py:76: in _pd_get_splits
    final_splits = tuple(
padl/util_transforms.py:77: in <genexpr>
    IfInStage(
padl/util_transforms.py:30: in __init__
    super().__init__(arguments=OrderedDict([('if_', if_),
padl/transforms.py:906: in __init__
    call = inspector.get_segment_from_frame(caller_frameinfo.frame, 'call')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

caller_frame = <frame at 0x7f9401eb26e0, file '/Users/sebastianlee/Aleph-One-Gmbh/padl/padl/util_transforms.py', line 77, code <genexpr>>, segment_type = 'call', return_locs = False

    def get_segment_from_frame(caller_frame: types.FrameType, segment_type, return_locs=False) -> str:
        """Get a segment of a given type from a frame.
    
        *NOTE*: All this is rather hacky and should be changed as soon as python 3.11 becomes widely
        available as then it will be possible to get column information from frames
        (see inline comments).
    
        *segement_type* can be 'call', 'attribute', 'getitem'.
        """
        if segment_type == 'call':
            node_type = ast.Call
            instructions_finder = _instructions_up_to_call
        elif segment_type == 'attribute':
            node_type = ast.Attribute
            instructions_finder = _instructions_in_name
        elif segment_type == 'getitem':
            node_type = ast.Subscript
            instructions_finder = _instructions_in_getitem
        # we want to extract the precise init statement here (e.g. `MyClass(1, 2, 3)`
        # , for python 3.11 (currently in development) this can be done via co_positions
        # (see https://www.python.org/dev/peps/pep-0657/),
        # for now, as 3.11 isn't widely used, this requires the following hack:
        # extract the source of the class init statement
        try:
            full_source = caller_frame.f_globals['_pd_source']
        except KeyError:
            full_source = get_source(caller_frame.f_code.co_filename)
    
        source, offset = get_statement(original(full_source), caller_frame.f_lineno)
        # the source can contain surrounding stuff we need to discard
        # as we only have the line number (this is what makes this complicated)
    
        # get all segments in the source that correspond to calls and might thus
        # potentially be the class init
        candidate_segments = var2mod.Finder(node_type).get_source_segments(source)
        # for each candidate, disassemble and compare the instructions to what we
        # actually have, a match means this is the correct statement
        if not candidate_segments:
            raise RuntimeError('No attributes found.')
        # disassemble and get the instructions up to the current position
        target_instrs = _instructions_up_to_offset(caller_frame.f_code,
                                                   caller_frame.f_lasti)
    
        segment = None
        locs = None
        found = False
    
        for segment, locs in sorted(candidate_segments, key=lambda x: -len(x[0])):
    
            instrs = instructions_finder(segment)
            if len(instrs) > len(target_instrs):
                continue
            for instr, target_instr in zip(instrs, target_instrs[-len(instrs):]):
                if (target_instr.opname == 'LOAD_FAST'
                        and target_instr.argval in caller_frame.f_locals
                        and target_instr.argval in caller_frame.f_code.co_varnames):
                    argval = caller_frame.f_locals[target_instr.argval]
                else:
                    argval = target_instr.argval
                if instr.argval != target_instr.argval and instr.argval != argval:
                    break
                same_opname = instr.opname == target_instr.opname
                load_ops = ('LOAD_NAME', 'LOAD_FAST', 'LOAD_GLOBAL', 'LOAD_CONST')
                both_load = instr.opname in load_ops and target_instr.opname in load_ops
                if not (same_opname or both_load):
                    break
            else:
                found = True
                break
    
        if segment is None or not found:
>           raise RuntimeError(f'{segment_type} not found.')
E           RuntimeError: call not found.

```

Reply:
I'll look into that. Seems like it's the same problem we had before with returning Identity (this happens if a Transform inside transforms.py creates another transfor), which I solved by introducing the BuiltinTransform. I'll check if there's some better solution to that.

Reply:
It doesn't seem to be a problem with the `CompoundTransform`'s